File: typescript/packages/core/src/commands/spec/parser.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { posixNormpath } from '../../workspace/expand/classify.ts'
import { AMBIGUOUS_NAMES } from './constants.ts'
import { type CommandSpec, OperandKind, ParsedArgs } from './types.ts'
import { rstripSlash } from '../../utils/slash.ts'

export function resolvePath(cwd: string, path: string): string {
  if (path.startsWith('/')) return posixNormpath(path)
  return posixNormpath(`${rstripSlash(cwd)}/${path}`)
}

function setValueFlag(
  flags: Record<string, string | boolean | string[]>,
  name: string,
  value: string,
  repeatFlags: ReadonlySet<string>,
): void {
  if (repeatFlags.has(name)) {
    const prev = flags[name]
    if (Array.isArray(prev)) {
      prev.push(value)
    } else {
      flags[name] = [value]
    }
  } else {
    flags[name] = value
  }
}

interface MixedCluster {
  bools: string[]
  valueFlag: string
  attached: string | null
}

// getopt-style cluster of bool flags ending in a value flag, e.g. -ne / -nepat.
// Returns null when any character is unknown or no value flag terminates it.
function matchMixedCluster(
  tok: string,
  boolFlags: ReadonlySet<string>,
  valueFlags: ReadonlySet<string>,
): MixedCluster | null {
  const bools: string[] = []
  const chars = tok.slice(1)
  for (let idx = 0; idx < chars.length; idx++) {
    const ch = chars[idx]
    if (ch === undefined) break
    const name = `-${ch}`
    if (boolFlags.has(name)) {
      bools.push(name)
      continue
    }
    if (valueFlags.has(name)) {
      const rest = chars.slice(idx + 1)
      return { bools, valueFlag: name, attached: rest.length > 0 ? rest : null }
    }
    return null
  }
  return null
}

export function parseCommand(spec: CommandSpec, argv: string[], cwd: string): ParsedArgs {
  const boolFlags = new Set<string>()
  const valueFlags = new Set<string>()
  const longBoolFlags = new Set<string>()
  const longValueFlags = new Set<string>()
  const valueFlagKinds = new Map<string, OperandKind>()
  const repeatFlags = new Set<string>()
  let numericShorthandFlag: string | null = null

  for (const opt of spec.options) {
    if (opt.short !== null) {
      if (opt.valueKind === OperandKind.NONE) {
        boolFlags.add(opt.short)
      } else {
        valueFlags.add(opt.short)
        valueFlagKinds.set(opt.short, opt.valueKind)
        if (opt.repeatable) repeatFlags.add(opt.short)
        if (opt.numericShorthand) numericShorthandFlag = opt.short
      }
    }
    if (opt.long !== null) {
      if (opt.valueKind === OperandKind.NONE) {
        longBoolFlags.add(opt.long)
      } else {
        longValueFlags.add(opt.long)
        valueFlagKinds.set(opt.long, opt.valueKind)
        if (opt.repeatable) repeatFlags.add(opt.long)
      }
    }
  }

  const restKind: OperandKind | null = spec.rest !== null ? spec.rest.kind : null

  const cachePaths: string[] = []
  const filteredArgv: string[] = []
  let i = 0
  while (i < argv.length) {
    const cur = argv[i]
    if (cur === '--cache') {
      i += 1
      for (;;) {
        const next = argv[i]
        if (next === undefined || next.startsWith('-')) break
        cachePaths.push(resolvePath(cwd, next))
        i += 1
      }
    } else {
      if (cur !== undefined) filteredArgv.push(cur)
      i += 1
    }
  }

  const flags: Record<string, string | boolean | string[]> = {}
  const rawArgs: string[] = []
  const warnings: string[] = []
  // Free-text commands (echo/python/bash-style TEXT rest) keep unknown dash
  // tokens verbatim; elsewhere they are dropped with a warning so a stray
  // flag never corrupts pattern/path classification.
  const lenientDashOperands = restKind === OperandKind.TEXT
  i = 0
  let endOfFlags = false

  while (i < filteredArgv.length) {
    const tok = filteredArgv[i]
    if (tok === undefined) break

    if (tok === '--' && !endOfFlags) {
      endOfFlags = true
      i += 1
      continue
    }

    if (endOfFlags) {
      rawArgs.push(tok)
      i += 1
      continue
    }

    if (tok.startsWith('--')) {
      if (longBoolFlags.has(tok)) {
        flags[tok] = true
        i += 1
      } else if (longValueFlags.has(tok) && i + 1 < filteredArgv.length) {
        setValueFlag(flags, tok, filteredArgv[i + 1] ?? '', repeatFlags)
        i += 2
      } else {
        const eq = tok.indexOf('=')
        if (eq !== -1 && longValueFlags.has(tok.slice(0, eq))) {
          setValueFlag(flags, tok.slice(0, eq), tok.slice(eq + 1), repeatFlags)
        } else if (lenientDashOperands) {
          rawArgs.push(tok)
        } else {
          warnings.push(`warning: unknown option '${tok}' ignored`)
        }
        i += 1
      }
      continue
    }

    if (tok.startsWith('-') && tok.length > 1) {
      if (numericShorthandFlag !== null && /^-\d+$/.test(tok)) {
        flags[numericShorthandFlag] = tok.slice(1)
        i += 1
        continue
      }
      let matchedValue = false
      for (const vf of valueFlags) {
        if (tok === vf && i + 1 < filteredArgv.length) {
          setValueFlag(flags, vf, filteredArgv[i + 1] ?? '', repeatFlags)
          i += 2
          matchedValue = true
          break
        }
        if (tok.startsWith(vf) && tok.length > vf.length) {
          setValueFlag(flags, vf, tok.slice(vf.length), repeatFlags)
          i += 1
          matchedValue = true
          break
        }
      }
      if (matchedValue) continue

      if (boolFlags.has(tok)) {
        flags[tok] = true
        i += 1
        continue
      }

      let allBool = true
      for (const ch of tok.slice(1)) {
        if (!boolFlags.has(`-${ch}`)) {
          allBool = false
          break
        }
      }
      if (allBool && tok.length > 1) {
        for (const ch of tok.slice(1)) flags[`-${ch}`] = true
        i += 1
        continue
      }

      const mixed = matchMixedCluster(tok, boolFlags, valueFlags)
      if (mixed !== null) {
        if (mixed.attached !== null) {
          for (const name of mixed.bools) flags[name] = true
          setValueFlag(flags, mixed.valueFlag, mixed.attached, repeatFlags)
          i += 1
          continue
        }
        if (i + 1 < filteredArgv.length) {
          for (const name of mixed.bools) flags[name] = true
          setValueFlag(flags, mixed.valueFlag, filteredArgv[i + 1] ?? '', repeatFlags)
          i += 2
          continue
        }
      }

      if (lenientDashOperands || /^-\d+$/.test(tok)) {
        rawArgs.push(tok)
      } else {
        warnings.push(`warning: unknown option '${tok}' ignored`)
      }
      i += 1
      continue
    }

    rawArgs.push(tok)
    i += 1
  }

  const positional: OperandKind[] = spec.positional
    .filter((op) => !op.providedBy.some((name) => name in flags))
    .map((op) => op.kind)

  const classified: [string, OperandKind][] = []
  const rawOperands: [string, OperandKind][] = []
  for (let j = 0; j < rawArgs.length; j++) {
    const arg = rawArgs[j]
    if (arg === undefined) continue
    let kind: OperandKind
    if (j < positional.length) {
      kind = positional[j] ?? OperandKind.TEXT
    } else if (restKind !== null) {
      kind = restKind
    } else {
      continue
    }
    if (kind === OperandKind.PATH) {
      classified.push([resolvePath(cwd, arg), OperandKind.PATH])
      rawOperands.push([arg, OperandKind.PATH])
    } else {
      classified.push([arg, OperandKind.TEXT])
      rawOperands.push([arg, OperandKind.TEXT])
    }
  }

  const pathFlagValues: string[] = []
  for (const [flagName, kind] of valueFlagKinds) {
    if (kind !== OperandKind.PATH || !(flagName in flags)) continue
    const val = flags[flagName]
    if (Array.isArray(val)) {
      const resolvedList = val.map((part) => resolvePath(cwd, part))
      flags[flagName] = resolvedList
      pathFlagValues.push(...resolvedList)
    } else if (typeof val === 'string') {
      const resolved = resolvePath(cwd, val)
      flags[flagName] = resolved
      pathFlagValues.push(resolved)
    }
  }

  const textFlagValues: string[] = []
  for (const [flagName, kind] of valueFlagKinds) {
    if (kind !== OperandKind.TEXT || !(flagName in flags)) continue
    const val = flags[flagName]
    if (Array.isArray(val)) {
      textFlagValues.push(...val)
    } else if (typeof val === 'string') {
      textFlagValues.push(val)
    }
  }

  return new ParsedArgs({
    flags,
    args: classified,
    cachePaths,
    pathFlagValues,
    rawOperands,
    textFlagValues,
    warnings,
  })
}

export function parseToKwargs(parsed: ParsedArgs): Record<string, string | boolean | string[]> {
  const result: Record<string, string | boolean | string[]> = {}
  for (const [key, value] of Object.entries(parsed.flags)) {
    let clean = key.replace(/^-+/, '').replaceAll('-', '_')
    clean = AMBIGUOUS_NAMES[clean] ?? clean
    result[clean] = value
  }
  return result
}

File: python/mirage/commands/spec/parser.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import posixpath
import re

from mirage.commands.spec.constants import AMBIGUOUS_NAMES
from mirage.commands.spec.types import CommandSpec, OperandKind, ParsedArgs

_NUMERIC_SHORT = re.compile(r"^-\d+$")


def _resolve(cwd: str, path: str) -> str:
    if path.startswith("/"):
        return posixpath.normpath(path)
    return posixpath.normpath(posixpath.join(cwd, path))


def _set_value_flag(
    flags: dict[str, str | bool | list[str]],
    name: str,
    value: str,
    repeat_flags: set[str],
) -> None:
    if name in repeat_flags:
        prev = flags.get(name)
        if isinstance(prev, list):
            prev.append(value)
        else:
            flags[name] = [value]
    else:
        flags[name] = value


def _match_mixed_cluster(
    tok: str,
    bool_flags: set[str],
    value_flags: set[str],
) -> tuple[list[str], str, str | None] | None:
    """Match a getopt-style cluster of bool flags ending in a value flag.

    Args:
        tok (str): token like "-ne" or "-nepat".
        bool_flags (set[str]): single-dash boolean flag names.
        value_flags (set[str]): single-dash value flag names.

    Returns:
        tuple[list[str], str, str | None] | None: (bool flag names, value
            flag name, attached value or None when the value comes from the
            next token), or None when any character is unknown or no value
            flag terminates the cluster.
    """
    bools: list[str] = []
    chars = tok[1:]
    for idx, ch in enumerate(chars):
        name = f"-{ch}"
        if name in bool_flags:
            bools.append(name)
            continue
        if name in value_flags:
            rest = chars[idx + 1:]
            return bools, name, (rest if rest else None)
        return None
    return None


def parse_command(
    spec: CommandSpec,
    argv: list[str],
    cwd: str,
) -> ParsedArgs:
    bool_flags: set[str] = set()
    value_flags: set[str] = set()
    long_bool_flags: set[str] = set()
    long_value_flags: set[str] = set()
    value_flag_kinds: dict[str, OperandKind] = {}
    repeat_flags: set[str] = set()
    numeric_shorthand_flag: str | None = None
    for opt in spec.options:
        if opt.short:
            if opt.value_kind == OperandKind.NONE:
                bool_flags.add(opt.short)
            else:
                value_flags.add(opt.short)
                value_flag_kinds[opt.short] = opt.value_kind
                if opt.repeatable:
                    repeat_flags.add(opt.short)
                if opt.numeric_shorthand:
                    numeric_shorthand_flag = opt.short
        if opt.long:
            if opt.value_kind == OperandKind.NONE:
                long_bool_flags.add(opt.long)
            else:
                long_value_flags.add(opt.long)
                value_flag_kinds[opt.long] = opt.value_kind
                if opt.repeatable:
                    repeat_flags.add(opt.long)

    rest_kind: OperandKind | None = (spec.rest.kind
                                     if spec.rest is not None else None)

    cache_paths: list[str] = []
    filtered_argv: list[str] = []
    i = 0
    while i < len(argv):
        if argv[i] == "--cache":
            i += 1
            while i < len(argv) and not argv[i].startswith("-"):
                cache_paths.append(_resolve(cwd, argv[i]))
                i += 1
        else:
            filtered_argv.append(argv[i])
            i += 1

    flags: dict[str, str | bool | list[str]] = {}
    raw_args: list[str] = []
    warnings: list[str] = []
    # Free-text commands (echo/python/bash-style TEXT rest) keep unknown
    # dash tokens verbatim; elsewhere they are dropped with a warning so a
    # stray flag never corrupts pattern/path classification.
    lenient_dash_operands = rest_kind == OperandKind.TEXT
    i = 0
    end_of_flags = False

    while i < len(filtered_argv):
        tok = filtered_argv[i]

        if tok == "--" and not end_of_flags:
            end_of_flags = True
            i += 1
            continue

        if end_of_flags:
            raw_args.append(tok)
            i += 1
            continue

        if tok.startswith("--"):
            if tok in long_bool_flags:
                flags[tok] = True
                i += 1
            elif tok in long_value_flags and i + 1 < len(filtered_argv):
                _set_value_flag(flags, tok, filtered_argv[i + 1], repeat_flags)
                i += 2
            else:
                eq = tok.find("=")
                if eq != -1 and tok[:eq] in long_value_flags:
                    _set_value_flag(flags, tok[:eq], tok[eq + 1:],
                                    repeat_flags)
                elif lenient_dash_operands:
                    raw_args.append(tok)
                else:
                    warnings.append(f"warning: unknown option '{tok}' ignored")
                i += 1
            continue

        if tok.startswith("-") and len(tok) > 1:
            if numeric_shorthand_flag is not None and _NUMERIC_SHORT.match(
                    tok):
                flags[numeric_shorthand_flag] = tok[1:]
                i += 1
                continue
            matched_value = False
            for vf in value_flags:
                if tok == vf and i + 1 < len(filtered_argv):
                    _set_value_flag(flags, vf, filtered_argv[i + 1],
                                    repeat_flags)
                    i += 2
                    matched_value = True
                    break
                if tok.startswith(vf) and len(tok) > len(vf):
                    _set_value_flag(flags, vf, tok[len(vf):], repeat_flags)
                    i += 1
                    matched_value = True
                    break
            if matched_value:
                continue

            if tok in bool_flags:
                flags[tok] = True
                i += 1
                continue

            all_bool = True
            for ch in tok[1:]:
                if f"-{ch}" not in bool_flags:
                    all_bool = False
                    break
            if all_bool and len(tok) > 1:
                for ch in tok[1:]:
                    flags[f"-{ch}"] = True
                i += 1
                continue

            mixed = _match_mixed_cluster(tok, bool_flags, value_flags)
            if mixed is not None:
                cluster_bools, vflag, attached = mixed
                if attached is not None:
                    for name in cluster_bools:
                        flags[name] = True
                    _set_value_flag(flags, vflag, attached, repeat_flags)
                    i += 1
                    continue
                if i + 1 < len(filtered_argv):
                    for name in cluster_bools:
                        flags[name] = True
                    _set_value_flag(flags, vflag, filtered_argv[i + 1],
                                    repeat_flags)
                    i += 2
                    continue

            if lenient_dash_operands or _NUMERIC_SHORT.match(tok):
                raw_args.append(tok)
            else:
                warnings.append(f"warning: unknown option '{tok}' ignored")
            i += 1
            continue

        raw_args.append(tok)
        i += 1

    positional: tuple[OperandKind,
                      ...] = tuple(op.kind for op in spec.positional
                                   if not any(name in flags
                                              for name in op.provided_by))

    classified: list[tuple[str, OperandKind]] = []
    raw_operands: list[tuple[str, OperandKind]] = []
    for j, arg in enumerate(raw_args):
        if j < len(positional):
            kind = positional[j]
        elif rest_kind is not None:
            kind = rest_kind
        else:
            continue
        if kind == OperandKind.PATH:
            classified.append((_resolve(cwd, arg), OperandKind.PATH))
            raw_operands.append((arg, OperandKind.PATH))
        else:
            classified.append((arg, OperandKind.TEXT))
            raw_operands.append((arg, OperandKind.TEXT))

    path_flag_values: list[str] = []
    for flag_name, kind in value_flag_kinds.items():
        if kind != OperandKind.PATH or flag_name not in flags:
            continue
        value = flags[flag_name]
        if isinstance(value, list):
            resolved_list = [_resolve(cwd, part) for part in value]
            flags[flag_name] = resolved_list
            path_flag_values.extend(resolved_list)
        elif isinstance(value, str):
            resolved = _resolve(cwd, value)
            flags[flag_name] = resolved
            path_flag_values.append(resolved)

    text_flag_values: list[str] = []
    for flag_name, kind in value_flag_kinds.items():
        if kind != OperandKind.TEXT or flag_name not in flags:
            continue
        value = flags[flag_name]
        if isinstance(value, list):
            text_flag_values.extend(value)
        elif isinstance(value, str):
            text_flag_values.append(value)

    return ParsedArgs(
        flags=flags,
        args=classified,
        cache_paths=cache_paths,
        path_flag_values=path_flag_values,
        raw_operands=raw_operands,
        text_flag_values=text_flag_values,
        warnings=warnings,
    )


def flag_kwarg_name(flag: str) -> str:
    clean = flag.lstrip("-").replace("-", "_")
    return AMBIGUOUS_NAMES.get(clean, clean)


def parse_to_kwargs(parsed: ParsedArgs) -> dict[str, str | bool | list[str]]:
    result: dict[str, str | bool | list[str]] = {}
    for key, value in parsed.flags.items():
        result[flag_kwarg_name(key)] = value
    return result
